pyphare/pyphare_tests/test_pharesee/test_geometry.py

```diff
@@ -0,0 +1,528 @@
+                    self.assertEqual(act_box, exp_box)
+
+
+
+
+
+    def test_overlapped_fields_are_equal(self):
+
+        overlaps = hierarchy_overlaps(self.hierarchy)
+
+        for ilvl, lvl in enumerate(self.hierarchy.patch_levels):
+
+            for overlap in overlaps[ilvl]:
+
+                pd1, pd2 = overlap["pdatas"]
+                box      = overlap["box"]
+                offsets  = overlap["offset"]
+
+                self.assertEqual(pd1.quantity, pd2.quantity)
+
+                if pd1.quantity == 'field':
+
+                    # we need to transform the AMR overlap box, which is thus
+                    # (because AMR) common to both pd1 and pd2 into local index
+                    # boxes that will allow to slice the data
+
+                    # the patchData ghost box that serves as a reference box
+                    # to transfrom AMR to local indexes first needs to be
+                    # shifted by the overlap offset associated to it
+                    # this is because the overlap box has been calculated from
+                    # the intersection of possibly shifted patch data ghost boxes
+
+                    loc_b1 = boxm.amr_to_local(box, boxm.shift(pd1.ghost_box, offsets[0]))
+                    loc_b2 = boxm.amr_to_local(box, boxm.shift(pd2.ghost_box, offsets[1]))
+
+                    data1 = pd1.dataset
+                    data2 = pd2.dataset
+
+                    slice1 = data1[loc_b1.lower:loc_b1.upper+1]
+                    slice2 = data2[loc_b2.lower:loc_b2.upper+ 1]
+
+                    self.assertTrue(np.allclose(slice1, slice2, atol=1e-12))
+
+
+
+
+
+
+
+    def test_overlapped_particledatas_have_identical_particles(self):
+
+        from copy import copy
+        overlaps = hierarchy_overlaps(self.hierarchy)
+
+        for ilvl, lvl in enumerate(self.hierarchy.patch_levels):
+
+            for overlap in overlaps[ilvl]:
+
+                pd1, pd2 = overlap["pdatas"]
+                box      = overlap["box"]
+                offsets  = overlap["offset"]
+
+                self.assertEqual(pd1.quantity, pd2.quantity)
+
+                if pd1.quantity == "particles":
+
+                    # the following uses 'offset', we need to remember that offset
+                    # is the quantity by which a patch has been moved to detect
+                    # overlap with the other one.
+                    # so shift by +offset when evaluating patch data in overlap box
+                    # index space, and by -offset when we want to shift box indexes
+                    # to the associated patch index space.
+
+                    # overlap box must be shifted by -offset to select data in the patches
+                    part1 = copy(pd1.dataset.select(boxm.shift(box, -offsets[0])))
+                    part2 = copy(pd2.dataset.select(boxm.shift(box, -offsets[1])))
+
+                    idx1 = np.argsort(part1.iCells + part1.deltas)
```

Comment:
not obvious to me : idx1 should be an index, while delta is not (it is a float or double) ?

Reply:
icell is the AMR index of the cell the particle is in.
deltas is the floating point position of the particle within the cell, normalised by mesh size (so between 0 and 1)

icell+delta is therefore the position of the particle, in "decimal integer" (you'd like to multiply by the mesh size to get x)

we sort by this key to be sure particles in the domain last cell and particle from the other patch ghost cell are sorted in the same order (we assume 'x' is a unique id of a particle since very improbable 2 particles get the same).

we need to sort particles if we want to compare them.